src/slide-image-source.cc

```diff
@@ -0,0 +1,164 @@
+    if (opts.scroll_animation ||
+        filename() == "/dev/stdin" || filename() == "-") {
+        return false;  // Not dealing with these now.
+    }
+
+    // Check if this format is supported
+    openslide_t *osr = openslide_open(filename().c_str());
+    if (!osr)
+        return false;
+
+    // Automatically cleanup openslide context
+    ScopeGuard s([osr]() { openslide_close(osr); });
+
+    std::unique_ptr<timg::Framebuffer> source_image;
+    int64_t width, height, l0_width, l0_height;
+    int target_width;
+    int target_height;
+
+    // We need to use one of the layers of the pyramid.
+    // Get native the highest resolution image.
+    openslide_get_level0_dimensions(osr, &l0_width, &l0_height);
+    if (invalid_dimensions(l0_width, l0_height))
+        return false;
+
+    // Get the target size
+    CalcScaleToFitDisplay(l0_width, l0_height, opts, false,
+                          &target_width, &target_height);
+
+    // First check if there's a thumbnail with enough resolution.
+    const char* const* associated_image_names = openslide_get_associated_image_names(osr);
+    while (*associated_image_names) {
+        const char *name = *associated_image_names;
+        openslide_get_associated_image_dimensions(osr, name, &width, &height);
+        if (invalid_dimensions(width, height))
+            return false;
+
+        if (strcmp(name, "thumbnail") == 0) {
+            // If the thumbnail width is smaller than the target,
+            // means we can sample from a larger image.
+            if (width < target_width)
+                break;
+
+            source_image.reset(new timg::Framebuffer(width, height));
+            openslide_read_associated_image(
+                osr, name, (uint32_t *) source_image->begin());
+            break;
+        }
+        associated_image_names++;
+    }
+
+    if (!source_image) {
+        // Get the best layer to downsample with scaling
+        // computed from layer 0 (the highest resolution one).
+        const double scaling = (double) l0_width / target_width;
+        const int32_t level = openslide_get_best_level_for_downsample(osr, scaling);
+        if (level < 0)
+            return false;
+
+        // Get target layer dimensions.
+        openslide_get_level_dimensions(osr, level, &width, &height);
+        if (invalid_dimensions(width, height))
+            return false;
+
+        source_image.reset(new timg::Framebuffer(width, height));
+        openslide_read_region(osr, (uint32_t *) source_image->begin(), 0, 0, level, width, height);
+        const char *err = openslide_get_error(osr);
+        if (err)
+            return false;
+    }
+
+    // TODO: Unpremultiply
+    // See: https://github.com/openslide/openslide/wiki/PremultipliedARGB
+
+    // Further scaling to desired target width/height
+    av_log_set_callback(dummy_log);
+    SwsContext *swsCtx = sws_getContext(width, height,
+                                        AV_PIX_FMT_RGB32,
```

Comment:
so `RGB32` is essentially ignoring 'A' ? Or does it magically make it through (as you showed in the images).
Can we change this to RGBA and see if it now works after we added the AlphaComposeBackground ?

Reply:
For reference, this is how it's defined:
https://ffmpeg.org/doxygen/3.4/pixfmt_8h.html#afa47a1e5163830c51665a3da04a318a0

As you can see it switches between the two formats depending on the endianess.